### PR TITLE
refactoring mate config

### PR DESCRIPTION
**Pull request by ideahitme**
1. Create config.go to store all mate configurations 
2. Remove flags initialisation from all internal packages (this is generally a bad practice to do so) 
3. Make internal packages more independent and actually "invocable"

### CODE REVIEW

**Comment by ideahitme** [comment]
@linki I would like to hear your opinion on this :)

File: config.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	FakeDNSName       string
+	FakeMode          string
+	FakeTargetDomain  string
+	FakeFixedDNSName  string
+	FakeFixedIP       string
+	FakeFixedHostname string
+
+	KubeServer     *url.URL
+	KubeFormat     string
+	TrackNodePorts bool
```
Comment (linki) [inline config.go]: how about - for consistency with the flags - to prefix with `kubernetes` or change the flags to `kube`? (the latter would require a version bump)

File: main.go [hunk 2]
```diff
@@ -1,50 +1,38 @@
...
 	}
 
-	p, err := producers.New(params.producer)
+	p, err := newProducer(cfg)
 	if err != nil {
 		log.Fatalf("Error creating producer: %v", err)
 	}
 
-	c, err := consumers.NewSynced(params.consumer)
+	c, err := newSyncConsumer(cfg)
```
Comment (linki) [inline main.go]: ~~use `newSyncedConsumer`~~

Comment (linki) [reply main.go]: or `newSynchronizedConsumer `

File: main.go [hunk 3]
```diff
@@ -57,3 +45,47 @@ func main() {
...
+	// New returns a Consumer implementation.
+	var consumer consumers.Consumer
+	var err error
+	switch cfg.Consumer {
+	case "google":
+		consumer, err = consumers.NewGoogleDNS(cfg.GoogleZone, cfg.GoogleProject, cfg.GoogleRecordGroupID)
+	case "aws":
+		consumer, err = consumers.NewAWSConsumer(cfg.AWSRecordGroupID)
+	case "stdout":
+		consumer, err = consumers.NewStdout()
```
Comment (linki) [inline main.go]: if you want we can also change these functions to:
* `consumers.NewGoogleCloudDNSConsumer`
* `consumers.NewAWSRoute53Consumer`
* `consumers.NewStdoutConsumer`
* `consumers.NewSynchronizedConsumer`

to have unified and unambiguous naming.

File: main.go [hunk 4]
```diff
@@ -57,3 +45,47 @@ func main() {
...
+			DNSName:       cfg.FakeDNSName,
+			FixedDNSName:  cfg.FakeFixedDNSName,
+			FixedHostname: cfg.FakeFixedHostname,
+			FixedIP:       cfg.FakeFixedIP,
+			Mode:          cfg.FakeMode,
+			TargetDomain:  cfg.FakeTargetDomain,
+		}
+		return producers.NewFake(fakeConfig)
+	case "null":
+		return producers.NewNull()
```
Comment (linki) [inline main.go]: let's change this to
* `producers.NewKubernetesProducer`
* `producers.NewFakeProducer`
* `producers.NewNullProducer`

File: producers/fake.go [hunk 5]
```diff
@@ -104,16 +104,16 @@ func (a *fakeProducer) generateEndpoint() (*pkg.Endpoint, error) {
...
 			byte(randomNumber(1, 255)),
 			byte(randomNumber(1, 255)),
 			byte(randomNumber(1, 255)),
 			byte(randomNumber(1, 255)),
 		).String()
-	case hostnameMode:
+	case "hostname":
 		endpoint.Hostname = fmt.Sprintf("%s.%s", randomString(6), a.targetDomain)
-	case fixedMode:
+	case "fixed":
```
Comment (linki) [inline producers/fake.go]: you still have those modes defined as `const` up there. We should use them. Also I think it's a good idea.

**Comment by linki** [comment]
I think this is a very valuable PR for mate 👍  I would be glad if you can sneak in the changes I proposed as well.

File: config.go (hunk 1 above)
Comment (ideahitme) [reply config.go]: +1 for kubernetes prefix :)

File: producers/fake.go (hunk 5 above)
Comment (ideahitme) [reply producers/fake.go]: yup, adding back

**Comment by ideahitme** [comment]
@linki thanks for valuable comments :) please take a look again

**Comment by linki** [comment]
awesome @ideahitme

**Comment by linki** [comment]
one last thing...

File: consumers/synced.go [hunk 6]
```diff
@@ -15,11 +15,7 @@ type SyncedConsumer struct {
 // one operation at a time, and blocks
 // concurrent operations until the current one
 // finishes.
-func NewSynced(name string) (Consumer, error) {
-	consumer, err := New(name)
-	if err != nil {
-		return nil, err
-	}
+func NewSynchronizedConsumer(consumer Consumer) (Consumer, error) {
 	return &SyncedConsumer{Consumer: consumer}, nil
```
Comment (linki) [inline consumers/synced.go]: let's also change this struct name

Comment (ideahitme) [reply consumers/synced.go]: @linki done :)

**Comment by linki** [comment]
👍

**Pull request merged**